File: 4 - Class vs. Instance/Solution.cpp
#include <iostream>

using namespace std;

class Person {
public:
    int age;

    Person(int initialAge);

    void amIOld();

    void yearPasses();
};

Person::Person(int initialAge) {
    // Add some more code to run some checks on initialAge
    if (initialAge > 0) age = initialAge;
    else {
        cout << "Age is not valid, setting age to 0." << endl;
        age = 0;
    }
}

void Person::amIOld() {
    // Do some computations in here and print out the correct statement to the console
    if (age < 13) cout << "You are young." << endl;
    else if (age < 18) cout << "You are a teenager." << endl;
    else cout << "You are old." << endl;
}

void Person::yearPasses() {
    // Increment the age of the person in here
    age++;
}

int main() {
    int t;
    int age;
    cin >> t;
    for (int i = 0; i < t; i++) {
        cin >> age;
        Person p(age);
        p.amIOld();
        for (int j = 0; j < 3; j++) {
            p.yearPasses();
        }
        p.amIOld();
        cout << '\n';
    }

    return 0;
}

File: 4 - Class vs. Instance/Solution.scala
class Person {

    var age: Int = 0

    def this(initialAge:Int) = {
        // Add some more code to run some checks on initialAge
        this()
        if (initialAge > 0) age = initialAge else {println("Age is not valid, setting age to 0."); age = 0}
            
    }        

    def amIOld(): Unit = {
       // Do some computations in here and print out the correct statement to the console 
       if (age < 13) println("You are young.")
           else if ((age >= 13) && (age < 18)) println("You are a teenager.")
           else println("You are old.")
    }

    def yearPasses(): Unit = {
        // Increment the age of the person in here
        age = age + 1
    }
}
